File: src/test/java/com/github/nsc/de/shake/parser/ParserTestUtil.java
package com.github.nsc.de.shake.parser;

import com.github.nsc.de.shake.lexer.Lexer;
import com.github.nsc.de.shake.lexer.characterinput.characterinputstream.CharacterInputStream;
import com.github.nsc.de.shake.lexer.characterinput.characterinputstream.SourceCharacterInputStream;
import com.github.nsc.de.shake.lexer.token.TokenInputStream;
import com.github.nsc.de.shake.parser.node.Node;
import com.github.nsc.de.shake.parser.node.Tree;

import static org.junit.jupiter.api.Assertions.*;
import static com.github.nsc.de.shake.TestUtil.*;

public class ParserTestUtil {

    public static Tree parse(String source, String input) {

        CharacterInputStream in = new SourceCharacterInputStream(source, input);
        Lexer lexer = new Lexer(in);
        TokenInputStream tokens = lexer.makeTokens();
        Parser parser = new Parser(tokens);

        return parser.parse();

    }

    @SuppressWarnings("unchecked")
    public static  <T extends Node> T parseSingle(String source, String input, Class<T> type) {

        Tree tree = parse(source, input);
        Node[] nodes = tree.getChildren();

        // Result should be 1 node
        assertEquals(1, nodes.length);

        // result is expected class
        assertNotNull(nodes[0]);
        assertType(type, nodes[0]);

        return (T) nodes[0];

    }
}

File: src/test/kotlin/com/github/nsc/de/shake/parser/ParserTestUtil.kt
package com.github.nsc.de.shake.parser

import com.github.nsc.de.shake.assertType
import com.github.nsc.de.shake.lexer.Lexer
import com.github.nsc.de.shake.parser.node.Node
import com.github.nsc.de.shake.parser.node.Tree
import com.github.nsc.de.shake.util.characterinput.characterinputstream.CharacterInputStream
import com.github.nsc.de.shake.util.characterinput.characterinputstream.SourceCharacterInputStream
import org.junit.jupiter.api.Assertions

object ParserTestUtil {
    fun parse(source: String?, input: String?): Tree {
        val input: CharacterInputStream = SourceCharacterInputStream(source, input)
        val lexer = Lexer(input)
        val tokens = lexer.makeTokens()
        val parser = Parser(tokens)
        return parser.parse()
    }

    @Suppress("unchecked")
    fun <T : Node?> parseSingle(source: String?, input: String?, type: Class<T>?): T {
        val tree = parse(source, input)
        val nodes = tree.children

        // Result should be 1 node
        Assertions.assertEquals(1, nodes.size)

        // result is expected class
        Assertions.assertNotNull(nodes[0])
        assertType(type, nodes[0])
        return nodes[0] as T
    }
}